---
title: "Multimodal Workflow Tutorial"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)
knitr::opts_chunk$set(echo = FALSE)
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v75)
library(ggplot2)
library(patchwork)
library(hdf5r)
#Create Data directory

#dir.create(paste(getwd(),"/data/",sep = ""))
options(timeout = 300)
# Get urls
raw_data_url <- "https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"
metadata_url <- "https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_singlecell.csv"
fragment_url <- "https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_fragments.tsv.gz"
fragment_index_url <- "https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_fragments.tsv.gz.tbi"
scRNAseq_url <-"https://signac-objects.s3.amazonaws.com/pbmc_10k_v3.rds"
# Get Destination
raw_data_dest <-  paste(getwd(), "/data/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5",sep="")
metadata_dest <-  paste(getwd(), "/data/atac_v1_pbmc_10k_singlecell.csv",sep="")
fragment_dest <- paste(getwd(), "/data/atac_v1_pbmc_10k_fragments.tsv.gz",sep="")
fragment_index_dest <- paste(getwd(), "/data/atac_v1_pbmc_10k_fragments.tsv.gz.tbi",sep="")
scrna_seq_dest <- paste(getwd(),"pbmc_10k_v3.rds",sep="")
Download file and add to directory
download.file(raw_data_url,raw_data_dest)
download.file(metadata_url,metadata_dest)
download.file(fragment_url,fragment_dest,mode="wb")
download.file(fragment_index_url,fragment_index_dest,mode="wb")
download.file(scRNAseq_url,scrna_seq_dest,mode="wb")

counts <- Read10X_h5(filename = raw_data_dest)
metadata <- read.csv(
  file = metadata_dest,
  header = TRUE,
  row.names = 1
)
chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = 'hg19',
  fragments =fragment_dest,
  min.cells = 10,
  min.features = 200
)
pbmc <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks",
  meta.data = metadata
)



tutorial_options(exercise.timelimit = 10000)
```

## Introduction

### **Multimodal Workflow Tutorial Combining scATAC-seq with scRNA-seq**
#### **Team BiocSwirl:** Sayan Bakshi, Yong Wang, Kingsley Bentum, Lisa N. Cao, Almas Khan (Team Leader)
##### This interactive course is adapted from PBMC scATAC-seq tutorial by Seurat/Signac team (https://satijalab.org/signac/articles/pbmc_vignette.html)

### What is ATAC-seq?

-   ATAC-seq stands for Assay for Transposase-Accessible Chromatin with high-throughput sequencing. 
-   The technique uses the Tn5 transposase enzyme to fragment chromatin while simultaneously integrating Next Generation Sequencing (NGS) adapters into the open (active) loci of chromatin. 
-   The library generated from the fragmentation is then sequenced by NGS and the open chromatin regions allowing for successful NGS adapters integration analyzed using bioinformatics tools. 
-   ATAC-seq is a simple two-step protocol compared to other techniques and requires a relatively low number of cells for the assay. 
-   With ATAC-seq, researchers are provided with information on chromatin accessibility across the entire genome and can be used as a screening tool even before further experimental works.

![](https://www.youtube.com/watch?v=L2Kxaq9yRE4)

```{r text_box, echo = FALSE}
question_text(
  "What is the full meaning of the acronym ATAC in 'ATAC-seq'? (Capitalize A-T-A-C):",
  answer("Assay for Transposase-Accessible Chromatin", correct = TRUE),
  allow_retry = TRUE
)
```

```{r quiz-0_1}
quiz(
  question("Which regions on a chromatin are considered active?",
    answer("open regions", correct = TRUE),
    answer("closed regions"),
    answer("wide regions"),
    answer("no regions"))
  )
```

```{r quiz-0_2}
quiz(
  question("The genes controlled by promoters and enhancers in an opened region will be expressed. (TRUE/FALSE)",
    answer("True", correct = TRUE),
    answer("False"))
  )
```

```{r quiz-0_3}
quiz(
  question("What is the name of the enzyme used in ATAC-seq?",
    answer("Taq polymerase"),
    answer("RNA polymerase"),
    answer("Tn5 transposase", correct = TRUE),
    answer("Reverse transcriptase"))
  )
```

```{r quiz-0_4}
quiz(
  question("Some information we can deduce from an ATAC-seq experiment include:",
    answer("Total number of genes on a chromosome"),
    answer("The opening of a chromatin loci", correct = TRUE),
    answer("The closure of chromatin loci", correct = TRUE),
    answer("All of the above"))
  )
```

```{r quiz-0_5}
quiz(
  question("An application of ATAC-seq in disease research may include:",
    answer("Identifying the ideal inhibitor to an undesiring gene expression"),
    answer("Comparing epigenetic markers in disease and normal states"),
    answer("Providing supporting data to RNA-seq and DNA methylation results"),
    answer("All of the above", correct = TRUE))
  )
```

Single-cell ATAC-seq (scATAC-seq) technology has also been developed to study cell type-specific chro-
matin accessibility in tissue samples containing a heterogeneous cellular population _(Baek S, Lee I. CSBJ. 2020)_. 

## **Topic 1 : Setting Up**

### Prerequisite
_R version > 4.1.3_ 

### Libraries

In this section, we will load the necessary packages and download the raw data.

Write the R code required to load the following necessary libraries:
_Signac, Seurat, GenomeInfoDb, EnsDb.Hsapiens.v75, ggplot2, patchwork, hdf5r_

```{r libraries-seeds, include=FALSE}
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v75)
library(ggplot2)
library(patchwork)
  library('hdf5r') ## was not included in the tutorial
set.seed(1234)
```

```{r load-libraries, exercise=TRUE, exercise.lines = 7}






```

### Setting seed

Next, we will set a random number as a seed to ensure that we get the same result everytime.

Write the R code to set a random seed:

```{r set-seed, exercise=T}

```

```{r quiz-1}
quiz(
  question("What is the purpose of setting seed?",
    answer("loading libraries"),
    answer("downloading a dataset"),
    answer("reproducibility", correct = TRUE))
  )
```

### Dataset

_The required data has been downloaded in the 'data' folder to be used in this tutorial_

## **Topic 2 : Pre-processing Workflow**

Signac uses peak/cell matrix and the fragment file information. To that end, we need to create a Seurat object using peak/cell matrix and cell metadata, and store the path to the fragment file on disk in the Seurat object.

```{r lib2, include=F}
library(sortable)
```

```{r sort-workflow, echo=FALSE}

steps <- c(
    "load filtered peak/cell matrix",
    "create chromatin assay with the genome information",
    "create a Seurat object to compute QC matrices"
    )

question_rank(
  "Sort these steps in the order of workflow:",
  answer(steps, correct = TRUE),
  allow_retry = TRUE
 )
```

### Load the peak matrix

```{r load-peak-matrix, exercise=TRUE, exercise.eval=TRUE}

```

```{r load-peak-matrix-hint}
counts <- Read10X_h5(filename = raw_data_dest)
```

### Load the metadata:

```{r load-metadata, exercise=TRUE, exercise.eval=TRUE}

```

```{r load-metadata-hint}
metadata <- read.csv(
  file = metadata_dest,
  header = TRUE,
  row.names = 1
)
```

### Create Chromatin Assay:

```{r chrom-assay, exercise=TRUE, exercise.eval=TRUE}

```

```{r chrom-assay-hint}
chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = 'hg19',
  fragments =fragment_dest,
  min.cells = 10,
  min.features = 200
)
```

```{r quiz-2}
quiz(
  question("Counts used in the creation of the ChromatinAssay are:",
    answer("Unnormalized or Raw", correct = TRUE),
    answer("Normalized"))
  )
```

Create a ChromatinAssay object to include features detected in at least 20 cells:

```{r chrom-assay-2, exercise=TRUE, exercise.eval=TRUE}

```

```{r chrom-assay-2-hint}
# change the value of min.cells to 20
```

### Create Seurat Object:

```{r seu-obj, exercise=TRUE, exercise.eval=TRUE}

```

```{r seu-obj-hint}
pbmc <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks",
  meta.data = metadata
)
```

Run a code to explore the stored peak feature and cell information within the SeuratObject:

```{r peaks-info}

```

```{r peaks-original, include=FALSE}

```

### Extract Gene Annotation and add it to the SeuratObject:

```{r gene-annot, exercise=TRUE, exercise.eval=TRUE}

```

```{r gene-annot-hint}
# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)

# change to UCSC style since the data was mapped to hg19
seqlevelsStyle(annotations) <- 'UCSC'

# add the gene information to the object
Annotation(pbmc) <- annotations

```

Run a code to include annotations from only protein-coding biotypes:

```{r annot-info, exercise=TRUE, exercise.eval=TRUE}

```

```{r annot-info-hint}
GetGRangesFromEnsDb(
  ensdb,
  standard.chromosomes = TRUE,
  biotypes = c("protein_coding"), # make sure that biotypes option only include 'protein-coding'
  verbose = TRUE
)
```

## **Topic 3 : Computing QC Metrices**

### Computing QC Metrics

Run a code to compute nucleosome signal score per cell:

```{r nucl-sig, exercise=TRUE, exercise.eval=TRUE}

```

```{r nucl-sig-hint}
pbmc <- NucleosomeSignal(object = pbmc)
```

```{r quiz-3, echo=FALSE}
quiz(
  question("Why we calculate NucleosomeSignal? (Select all that apply)",
  answer("Quantify the approximate ratio of mononucleosomal to nucleosome-free fragments", correct = TRUE),
  answer("The histogram of DNA fragment sizes should exhibit a strong nucleosome banding pattern", correct = TRUE))
)
```

Next, we will compute TSS (Transcription Start Site) enrichment score per cell with `TSSEnrichment`. It takes 15-20 minutes to run. You can allow it to run or alternatively, load a pre-run object at this step.

```{r tss-enrich, include=FALSE, eval=FALSE}
pbmc <- TSSEnrichment(object = pbmc, fast = FALSE)
```

Calculate the blacklist ratio and fraction of reads in peaks and add it to the SeuratObject:

```{r blacklist, exercise=TRUE, exercise.eval=TRUE}

```

```{r blacklist-hint}
pbmc$pct_reads_in_peaks <- pbmc$peak_region_fragments / pbmc$passed_filters * 100
pbmc$blacklist_ratio <- pbmc$blacklist_region_fragments / pbmc$peak_region_fragments
```

```{r quiz-4, echo=FALSE}
quiz(
  question("What are the QC metrics for the scATAC-seq experiment? (Select all that apply)",
  answer("Nucleosome banding pattern", correct = TRUE),
  answer("Transcriptional start site (TSS) enrichment score", correct = TRUE),
  answer("Total number of fragments in peaks", correct = TRUE),
  answer("Fraction of fragments in peaks:", correct = TRUE),
  answer("Ratio reads in genomic blacklist regions ", correct = TRUE))
)
```

In the next step, inspect the TSS enrichment scores by grouping the cells based on the score:

```{r tss-filter, exercise=TRUE, exercise.eval=TRUE}

```

```{r tss-filter-hint}
pbmc$high.tss <- ifelse(pbmc$TSS.enrichment > 2, 'High', 'Low')
```

Plot the accessibility signal over all TSS sites with `TSSPlot` by grouping cells with high or low nucleosomal signal strength:

```{r tss-filter-plot, exercise=TRUE, exercise.eval=TRUE}

```

```{r tss-filter-plot-hint}
TSSPlot(pbmc, group.by = 'high.tss') + NoLegend()
```

Inspect the fragment length periodicity for all the cells and group by cells with high or low nucleosomal signal strength:

```{r frag-hist, exercise=TRUE, exercise.eval=TRUE}

```

```{r frag-hist-hint}
pbmc$nucleosome_group <- ifelse(pbmc$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
FragmentHistogram(object = pbmc, group.by = 'nucleosome_group')
```

Make violin plots for QC metrics:
_QC metrics to include: 'pct_reads_in_peaks', 'peak_region_fragments', 'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'_

```{r vlnplot, exercise=TRUE, exercise.eval=TRUE}

```

```{r vlnplot-hint}
VlnPlot(
  object = pbmc,
  features = c('pct_reads_in_peaks', 'peak_region_fragments',
               'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 5
)
```

Remove cells that are outliers for QC metrices:

```{r filter-cell, exercise=TRUE, exercise.eval=TRUE}

```

```{r filter-cell-hint}
pbmc <- subset(
  x = pbmc,
  subset = peak_region_fragments > 3000 &
    peak_region_fragments < 20000 &
    pct_reads_in_peaks > 15 &
    blacklist_ratio < 0.05 &
    nucleosome_signal < 4 &
    TSS.enrichment > 2
)
```

## **Topic 4 : Normalization and linear dimensional reduction**

We will use the combined steps of TF-IDF followed by SVD (known as LSI), which were first introduced for the analysis of scATAC-seq data by _Cusanovich et al. 2015_.

```{r norm, exercise=TRUE, exercise.eval=TRUE}

```

```{r norm-hint}
pbmc <- RunTFIDF(pbmc)
pbmc <- FindTopFeatures(pbmc, min.cutoff = 'q0')
pbmc <- RunSVD(pbmc)
```

```{r quiz-5, echo=FALSE}
quiz(
  question("What is the two-step normalization precedure of Signac? (Select all that apply)",
  answer("Normalize across cells to correct for differences in cellular sequencing depth", correct =TRUE),
  answer("Normalize across peaks to give higher values to more rare peaks", correct = TRUE))
)  
```

```{r quiz-6, echo=FALSE}
quiz(question("What type of normalization does Signac perform? (Select all that apply)",
  answer("Term frequency-inverse document frequency (TF-IDF)", correct =TRUE),
  answer("Normalize across peaks to give higher values to more rare peaks", correct = TRUE)
))
```

```{r quiz-7, echo=FALSE}
quiz(question("What type of dimension reduction does Signac perform? (select All that apply)",
  answer("singular value decomposition (SVD), similar to the output of PCA used for scRNA-seq", correct =TRUE),
  answer("normalize across peaks to give higher values to more rare peaks", correct = TRUE)
))
```

```{r text_box_2, echo = FALSE}
question_text(
  "What are the combined steps of TF-IDF followed by SVD known as? (Capitalize):",
  answer("Latent Semantic Indexing", correct = TRUE),
  allow_retry = TRUE
)
```

## **Topic 5 : Non-linear dimension reduction and clustering**

Utilize common scRNAseq-data-analysis functions:

```{r quiz-8, echo=FALSE}
quiz(
  question("What are the fuctions used for non-linear dimension reduction and clustering?",
  answer("RunUMAP()", correct =TRUE),
  answer("FindNeighbors()", correct = TRUE),
  answer("FindCluters()", correct = TRUE),
  answer("SVD"),
  answer("PCA"))
)
```

```{r nldrc, exercise=TRUE, exercise.eval=TRUE}

```

```{r nldrc-hint}
pbmc <- RunUMAP(object = pbmc, reduction = 'lsi', dims = 2:30)
pbmc <- FindNeighbors(object = pbmc, reduction = 'lsi', dims = 2:30)
pbmc <- FindClusters(object = pbmc, verbose = FALSE, algorithm = 3)
```

Perform visualization of non-linear dimension reduction and graph-based clustering:

```{r nldrc-plot, exercise=TRUE, exercise.eval=TRUE}

```

```{r nldrc-plot-hint}
DimPlot(object = pbmc, label = TRUE) + NoLegend()
```

## **Topic 6 : Create a Gene Activity Matrix**

With UMAP visualization, you can identify the presence of multiple cell groups. Now, we will assess the chromatin accessibility of each gene to quantify its activity by creating a gene activity matrix (using `GeneActivity`) and adding it to the Seurat object.

```{r gam, exercise=TRUE, exercise.eval=TRUE}

```

```{r gam-hint}
gene.activities <- GeneActivity(pbmc)
pbmc[['RNA']] <- CreateAssayObject(counts = gene.activities)
```

Perform normalization on the new Seurat object with `NormalizeData`:
```{r norm-gam, exercise=TRUE, exercise.eval=TRUE}

```

```{r norm-gam-hint}
pbmc <- NormalizeData(
  object = pbmc,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(pbmc$nCount_RNA)
)
```

**Exercise**
Normalize the same object with centered log ratio transformation (use a different variable)

```{r norm-gam-ex, exercise=TRUE}

```

```{r quiz-9, echo=FALSE}
quiz(
  question("Why are the activities of canonical marker genes much noisier than scRNA-seq?",
  answer("They represent measurements from sparse chromatin data", correct =TRUE),
  answer("They include raw unnormalized data"),
  answer("They assume a general correspondence between gene body/promoter accessibility and gene expression which may not always be the case", correct = TRUE))
)
```

Create multiple plots identifying 4 different populations by their markers (MS4A1, CD3D, LEF1, NKG7)

```{r feat-plot, exercise=TRUE, exercise.eval=TRUE}

```

```{r feat-plot-hint}
DefaultAssay(pbmc) <- 'RNA'

FeaturePlot(
  object = pbmc,
  features = c('MS4A1', 'CD3D', 'LEF1', 'NKG7'),
  pt.size = 0.1,
  max.cutoff = 'q95',
  ncol = 3
)
```

## **Topic 7 : Integrating scATAC-seq data with scRNA-seq data**

While scATAC-seq informs us of chromatin accessibility changes, incorporating scRNA-seq data from the same system (human PBMC) provides quantitative gene expression data associated with differential chromatin accessibility changes surrounding respective TSS. In this method, we will utilize methods for cross-modality integration and label transfer (Cell. Volume 177, Issue 7, 13 June ## 2019, Pages 1888-1902.e21.). Hence, we can identify the shared changes across two modalities for relevant biological inference. 

### scRNA-seq Dataset
_The required data has been downloaded in the 'data' folder to be used in this tutorial_

### Workflow

**Step 1:** Load the pre-processed scRNA-seq data for PBMCs and find 'Anchors' that will help integrating both datasets using `FindTransferAnchors` and `TransferData`

```{r RnaData, exercise=TRUE, exercise.eval=TRUE}

```

```{r RnaData-hint}

pbmc_rna <- readRDS(scrna_seq_dest)

transfer.anchors <- FindTransferAnchors(
  reference = pbmc_rna,
  query = pbmc,
  reduction = 'cca'
)

predicted.labels <- TransferData(
  anchorset = transfer.anchors,
  refdata = pbmc_rna$celltype,
  weight.reduction = pbmc[['lsi']],
  dims = 2:30
)

```

**Step 2:** Add metadata information of predicted labels to the Seurat object

```{r RnaData_meta, exercise=TRUE, exercise.eval=TRUE}

```

```{r RnaData_meta-hint}
pbmc <- AddMetaData(object = pbmc, metadata = predicted.labels)
```

Create UMAP visualization with the scRNA-seq and scATAC-seq data:

```{r rna-atac-plot, exercise=TRUE, exercise.eval=TRUE}

```

```{r rna-atac-plot-hint}
plot1 <- DimPlot(
  object = pbmc_rna,
  group.by = 'celltype',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scRNA-seq')

plot2 <- DimPlot(
  object = pbmc,
  group.by = 'predicted.id',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scATAC-seq')

plot1 + plot2
```

These plots exhibit that scRNA-based classifications are consistent with the scATAC-seq based UMAP visualization.

```{r quiz-10, echo=FALSE}
quiz(question("Which plot shows the predicted labels?)",
  answer("Left (scRNA-seq)"),
  answer("Right (scATAC-seq)", correct = TRUE))
)  
```

```{r quiz-11, echo=FALSE}
quiz(question("Why cluster 14 should be removed from the downstream analysis?)",
  answer("This group has lower QC metrics", correct = TRUE),
  answer("There is no correlation between gene activity matrix and scRNA-seq dataset"))
)  
```

**Step 3:** Annotate the scATAC-seq-derived clusters

_Cluster 14 (mapped to CD4 Memory T cells) is an extremely small cluster with lower QC metrics, potentially representing low-quality cells. Hence, it should be removed from downstream analysis._

```{r annot-multi, exercise=TRUE, exercise.eval=TRUE}

```

```{r annot-multi-hint}
pbmc <- subset(pbmc, idents = 14, invert = TRUE)
pbmc <- RenameIdents(
  object = pbmc,
  '0' = 'CD14 Mono',
  '1' = 'CD4 Memory',
  '2' = 'CD8 Effector',
  '3' = 'CD4 Naive',
  '4' = 'CD14 Mono',
  '5' = 'DN T',
  '6' = 'CD8 Naive',
  '7' = 'NK CD56Dim',
  '8' = 'pre-B',
  '9' = 'CD16 Mono',
  '10' = 'pro-B',
  '11' = 'DC',
  '12' = 'NK CD56bright',
  '13' = 'pDC'
)
```

```{r quiz-12, echo=FALSE}
quiz(question("What description is correct for the following answer?",
  answer("Transfer the cell-type labels we learned from the scRNA-seq data to the scATAC-seq cells", correct = TRUE),
  answer("Transfer the scATAC-seq labels to the scRNA-seq celltype labels"))
)  
```

